test/test_symmetry.jl

```diff
@@ -213,63 +250,193 @@ end
+    end
+
+    # Internal conversion between spin and Stevens operators
+    let
+        J = randn(3,3)
+        J = J+J'
+        p = randn(3)'*𝒮 + 𝒮'*J*𝒮 +
+            randn(11)' * Sunny.stevens_operator_symbols[5] +
+            randn(13)' * Sunny.stevens_operator_symbols[6]
+        cp = Sunny.operator_to_classical_polynomial(p)
+        @test cp |> Sunny.classical_polynomial_to_classical_stevens |> Sunny.operator_to_classical_polynomial ≈ cp
     end
 
     # Test some inferred anisotropy matrices
-    begin
+    let
         N = 7
         k = 6
         i = 1
         cryst = Sunny.diamond_crystal()
 
-        # print_allowed_anisotropy(cryst, i; k)
-        𝒪 = stevens_operators(N, k)
-        Λ = 𝒪[0]-21𝒪[4]
+        # print_site(cryst, i)
+        Λ = 𝒪[6,0]-21𝒪[6,4]
         @test Sunny.is_anisotropy_valid(cryst, i, Λ)
 
-        R = hcat(normalize([1, 1, -2]), normalize([-1, 1, 0]), normalize([1, 1, 1]))
-        R = Sunny.Mat3(R)
-        # print_allowed_anisotropy(cryst, i; k, R, digits=14)
-        𝒪 = stevens_operators(N, k; R)
-        Λ = 𝒪[0]-12.37436867076458𝒪[3]+9.625𝒪[6]
-        @test Sunny.is_anisotropy_valid(cryst, i, Λ)
+        R = [normalize([1 1 -2]); normalize([-1 1 0]); normalize([1 1 1])]
+        # print_site(cryst, i; R)
+        Λ = 𝒪[6,0]-(35/√8)*𝒪[6,3]+(77/8)*𝒪[6,6]
+        Λ′ = rotate_operator(Λ, R)
+        @test Sunny.is_anisotropy_valid(cryst, i, Λ′)
 
         lat_vecs = lattice_vectors(1.0, 1.1, 1.0, 90, 90, 90)
         cryst = Crystal(lat_vecs, [[0., 0., 0.]])
-        # print_allowed_anisotropy(cryst, i; k, digits=14)
-        𝒪 = stevens_operators(N, k)
-        Λ = randn()*(𝒪[0]-21𝒪[4]) + randn()*(𝒪[2]+3.2𝒪[4]+2.2𝒪[6])
+        # print_site(cryst, i)
+        Λ = randn()*(𝒪[6,0]-21𝒪[6,4]) + randn()*(𝒪[6,2]+(16/5)*𝒪[6,4]+(11/5)*𝒪[6,6])
         @test Sunny.is_anisotropy_valid(cryst, i, Λ)
     end
-end
 
-@testitem "Rotations" begin
-    include("test_shared.jl")
+    # Test fast evaluation of Stevens operators
+    let
+        import DynamicPolynomials as DP
 
-    A = randn(3,3)
-    R = exp(A - A')
-    (n, θ) = Sunny.axis_angle(Sunny.Mat3(R))
-    @test 1 + 2cos(θ) ≈ tr(R)
-    @test norm(n) ≈ 1
-    @test R*n ≈ n
+        s = randn(Sunny.Vec3)
+        p = randn(5)' * Sunny.stevens_operator_symbols[2] + 
+            randn(9)' * Sunny.stevens_operator_symbols[4] +
+            randn(13)' * Sunny.stevens_operator_symbols[6]
+        (_, c2, _, c4, _, c6) = Sunny.operator_to_classical_stevens_coefficients(p, 1.0)
+
+        p_classical = Sunny.operator_to_classical_polynomial(p)
+        grad_p_classical = DP.differentiate(p_classical, Sunny.spin_classical_symbols)
+
+        E_ref = p_classical(Sunny.spin_classical_symbols => s)
+
+        gradE_ref = [g(Sunny.spin_classical_symbols => s) for g = grad_p_classical]
+        gradE_ref -= (gradE_ref⋅s)*s / (s⋅s) # Orthogonalize to s
```

Comment:
Recall discussion about this. Will probably need to ask you to explain more at a later point.

Reply:
We should have it documented. The question is: what should be the gradient of `X = |S|^2` with respect `S`? If you treat `X` as a scalar, the gradient is 0. If you treat `X` as a function of `S`, the gradient is `2S`. In practice, (e.g. for spin dynamics) we only use the part of the gradient that is _orthogonal_ to `S`, in which case it is always reasonable to treat `grad X = 0`. That makes it much easier to implement `energy_and_gradient_for_classical_anisotropy`.

Reply:
I added a comment to this effect in test_symmetry.jl